src/identity.cc

```diff
@@ -210,40 +210,37 @@ ModelState::ValidateModelConfig()
   RETURN_IF_ERROR(model_config_.MemberAsArray("input", &inputs));
   RETURN_IF_ERROR(model_config_.MemberAsArray("output", &outputs));
 
-  // There must be 1 input and 1 output.
+  // There must be equal number of inputs and outputs.
   RETURN_ERROR_IF_FALSE(
-      inputs.ArraySize() == 1, TRITONSERVER_ERROR_INVALID_ARG,
-      std::string("expected 1 input, got ") +
-          std::to_string(inputs.ArraySize()));
-  RETURN_ERROR_IF_FALSE(
-      outputs.ArraySize() == 1, TRITONSERVER_ERROR_INVALID_ARG,
-      std::string("expected 1 output, got ") +
-          std::to_string(outputs.ArraySize()));
-
-  common::TritonJson::Value input, output;
-  RETURN_IF_ERROR(inputs.IndexAsObject(0, &input));
-  RETURN_IF_ERROR(outputs.IndexAsObject(0, &output));
-
-  // Input and output must have same datatype
-  std::string input_dtype, output_dtype;
-  RETURN_IF_ERROR(input.MemberAsString("data_type", &input_dtype));
-  RETURN_IF_ERROR(output.MemberAsString("data_type", &output_dtype));
-
-  RETURN_ERROR_IF_FALSE(
-      input_dtype == output_dtype, TRITONSERVER_ERROR_INVALID_ARG,
-      std::string("expected input and output datatype to match, got ") +
-          input_dtype + " and " + output_dtype);
-
-  // Input and output must have same shape
-  std::vector<int64_t> input_shape, output_shape;
-  RETURN_IF_ERROR(backend::ParseShape(input, "dims", &input_shape));
-  RETURN_IF_ERROR(backend::ParseShape(output, "dims", &output_shape));
-
-  RETURN_ERROR_IF_FALSE(
-      input_shape == output_shape, TRITONSERVER_ERROR_INVALID_ARG,
-      std::string("expected input and output shape to match, got ") +
-          backend::ShapeToString(input_shape) + " and " +
-          backend::ShapeToString(output_shape));
+      inputs.ArraySize() == outputs.ArraySize(), TRITONSERVER_ERROR_INVALID_ARG,
+      std::string("model configuration must have equal input/output pairs"));
+
+  for (size_t io_index = 0; io_index < inputs.ArraySize(); io_index++) {
```

Comment:
Identity in multiple I/O case maps the input to output based on the naming (i.e. "INPUT0" to "OUTPUT0"). You shouldn't assume the matching input and output are at the same index, you may have a config like below while we actually want to validate the spec between "INPUT0" and "OUTPUT0"
```
input [
{ name: "INPUT2" ...},
{ name: "INPUT0" ...}, 
{ name: "INPUT1" ...}
]
output [
{ name: "OUTPUT0" ...},
{ name: "OUTPUT1" ...}, 
{ name: "OUTPUT2" ...}
]
```

Reply:
Have your new changes addressed this comment? What will your current change do to validate the config above?

Reply:
Fixed by latest commit